### ISSUE TITLE
Redux State Design

### ISSUE THREAD

**Issue by SirPython** [Cardshifter/HTML-Client#127]
Before we start rewriting the client, we should discuss how our state tree will look like. Things to take into consideration:
- State for/from:
  - Login
  - Lobby
  - Deck Builder
  - Game Screen
  - Server Interface

The state for the Server Interface has already been briefly discussed by @danpantry in #116

**Comment by SirPython** [comment]
Not about the specific design, but more about what each scene should handle:

- Login: user info, server info
- Lobby: chat messages, online users, available mods
- Deck Builder: available cards, user decks
- Game Screen: user deck, user battlefield, opponent battlefield, ?individual card state? (maybe that should be handled by individual component state), etc.
- Server Interface: handles connection between other scenes' business logic and the server

**Comment by danpantry** [comment]
> ?individual card state? (maybe that should be handled by individual component state)

What sort of state are we talking here? "Hovered/Selected"?

**Comment by Phrancis** [comment]
This is by all means partial, incomplete and tentative, but here are the things I thought of in regards to Redux for the client:

State for/from:

Login

- Servers (id, name, url, port, secureEnabled?, status?)
- userName
- selectedServer
- selectedIsSecure
- connection? (serverId, serverResponse, connectTimestamp?, disconnectTimestamp?)

Lobby

- ConnectedUsers (id, name, status?)
- Mods (id, name, ??)
- ChatMessages (msgId, playerId, msgText, msgTimestamp)
- Invites (id, inviterUserId, inviteeUserId, modId, inviteTimestamp, accepted?)
- Commands? (id, name, action)
- userId
- selectedMod 
- selectedUserId

Deck Builder

(Note: This assumes the existence of Card objects from server)
- AllCards (cardId, name, other props...)
- UserDecks (id, name, [deckCards (cardId, cardCount)], versionId?, versionTimestamp?)
- CurrentDeckCards (deckCards [cardId, cardCount])
- Commands (id, name, action)
- countOfCurrentDeckCards?
- selectedCard (cardId)

**Comment by SirPython** [comment]
@danpantry Well, I think cards can have health and such. Also, they may be under some sort of effect.

@Phrancis Nice! A few things though: I don't think the login scene should necessarily store `selectedServer` and `selectedIsSecure`. These are just form options and are most likely going to be used to immediately initialize the server interface. What are your thoughts on this?

Also, I don't think the commands for the lobby need to be stored in the redux tree. Shouldn't they just be stored on a local map someplace else? I feel like trying to push all the commands onto the tree just so one component/thing can use them is unnecessary. That is, unless the commands for Deck Builder are the same.

**Comment by Phrancis** [comment]
OK @SirPython makes sense, I edited my comment accordingly.

**Comment by SirPython** [comment]
This is some thought towards the server interface. This should be done with middleware.

*Note: the lists below are all convoluted.*

<hr>

Sending a message:

1. Component/container dispatches an event like this:
```
{
    type: "SEND_MESSAGE":
    payload: ... one of the types, as have already been defined in the current server interface...
}
```
2. Server interface takes in the payload, JSON.stringify's it, and sends it via the websocket to the server.

<hr>

Receiving a message.

1. The server interface is listening to the websocket for activity.
2. Once a message is received, the payload is JSON.parse'd and an action is dispatched:
```
{
    type: "MESSAGE_RECEIVED",
    payload: ... the parsed object...
}
```

3. Any components waiting for messages will see the action and will read the payload to see if they should deal with the message.

<hr>
<hr>

Alternate idea:

Rather than having two actions `SEND_MESSAGE` and `MESSAGE_RECEIVED`, perhaps each different type of action could have it's own type of action. For example, there'd be a `CHAT_MESSAGE`. That way, if possible, containers can filter out what they receive from the store based on the type. If that's not possible, than either way could go, I guess.

**Comment by danpantry** [comment]
I'd say we should use `MESSAGE_RECEIVED` and `SEND_MESSAGE`, but not those exact names. Have them be names exported from the middleware module we create and have *only the middleware* use those action names. Then have "secondary" action names that correspond to other actions. Example:

````
// socketMiddleware.js

export const MESSAGE_RECEIVED = '@@message-middleware/MESSAGE_RECEIVED';

const middlewareFactory = transformReceivedAction => store => next => action => {
  if (action.type === MESSAGE_RECEIVED) {
    const message = transformReceivedAction(action);
    return next(message);
  }
}
export default middlewareFactory;

// store.js
import socketMiddlewareFactory from './socketMiddleware';
import {createStore, applyMiddleware} from 'redux';
import reducer from './reducer';

const socketMiddleware = socketMiddlewareFactory(function onReceiveMessage(action) {
  const finalAction = {...action, ....}; // mutate the action somehow
  return finalAction;
}));
export default createStore(reducer, null, applyMiddleware(socketMiddleware));
````

This is all chicken scratch of course

**Comment by Phrancis** [comment]
Closing as no longer needed #144
